numpy/_core/_add_newdocs.py

```diff
@@ -5350,9 +5350,9 @@
     axis : int, optional
         The axis along which to apply the reduceat.
     dtype : data-type code, optional
-        The type used to represent the intermediate results. Defaults
-        to the data type of the output array if this is provided, or
-        the data type of the input array if no output array is provided.
+        The type used to perform the intermediate operations. Defaults to the
```

Comment:
I actually do not understand "intermediate results" here at all. Shouldn't it just be
```
The data type used to perform the operation. Defaults to that of ``out`` if given,
and of the inputs (``array`` and possible identity).
```

Reply:
Hmm, given my further thought in https://github.com/numpy/numpy/issues/25752#issuecomment-1924486036, I think this should be rephrased... Maybe,
```
The data type used to perform the operation. Defaults to that of ``out`` if given,
and the data type of ``array`` otherwise (though upcast to conserve precision
for some cases, such as `numpy.add.reduce` for integer or boolean input).
```

Reply:
Mentioning `identity` below is probably wrong.  `initial` could take part, but doesn't.  `out`, I am not 100% sure, it probably does take part in promotion (for reduction that isn't that weird, since you effectively do `out + val`).

Overall, reductions are rather strict though (otherwise they don't work), so double checking out, that sounds right.

Reply:
@eliegoudout - The text still talks about casting together with `initial`, which is not quite correct. Could you change to (something like) the text I wrote above, in https://github.com/numpy/numpy/pull/25753#discussion_r1476542069

Reply:
@mhvk I don't understand what you re referring to. In the current version (possibly final imo) of the PR, the `dtype` doc matches between `reduce` and `reduceat`, which is good imo since `initial` doesn't have any impact on casting
```pycon
>>> np.add.reduce([True, True], initial=.9)
2  # initial is casted to np.add.identity dtype.
```

The PR LGTM right now, but I might have overlooked something.

Reply:
Right now, you have "... or the result of `ufunc`'s casting of the input array with `ufunc.identity`." I don't think this is correct, it depends *only* on the input array (see @seberg's comment), but for `np.add` and `np.multiply` any integer, including bool, is upcast to `int64` (which is probably what caused the confusion -- since `type(np.add.identity) is int` -- but this is not actually how it is implemented). Hence, my suggestion to rephrase as in https://github.com/numpy/numpy/pull/25753/files#r1476542069 (but change `.reduce` to `.reduceat` for the `reduceat` docstring).

Just to show that the type of `identity` has no effect:
```
In [7]: np.add.reduceat([True, True], [0]).dtype
Out[7]: dtype('int64')

In [8]: np.minimum.reduceat([True, True], [0]).dtype
Out[8]: dtype('bool')

In [9]: np.floor_divide.reduceat([True, True], [0]).dtype
Out[9]: dtype('int8')
```

Reply:
> Right now, you have "... or the result of `ufunc`'s casting of the input array with `ufunc.identity`." I don't think this is correct, it depends _only_ on the input array

I think this is not correct:
```pycon
>>> out = np.zeros(())
>>> np.add.reduce([True, True], out=out)
array(2.)  # dtype is determined by `out`
```

> Just to show that the type of `identity` has no effect:
> 
> ```
> In [7]: np.add.reduceat([True, True], [0]).dtype
> Out[7]: dtype('int64')
> 
> In [8]: np.minimum.reduceat([True, True], [0]).dtype
> Out[8]: dtype('bool')
> 
> In [9]: np.floor_divide.reduceat([True, True], [0]).dtype
> Out[9]: dtype('int8')
> ```
> 
> EDIT: actually, not sure the above proves anything, since the identity for `minimum` and `floor_divide` is `None`. But I happen to know the implementation...

You examples tend to show that the current version of the PR is correct, since as you pointed out, when `ufunc.identity` is not `None`,  it seems to be the dtype determinant, **unless** `out` was provided.

In conclusion: I think the version I propose is correct.

Reply:
But your version is not how it is implemented! While mine is. And it might help people to understand why, e.g., the following happens:
```
In [2]: np.subtract.reduce(np.array([1, 2], dtype='i1')).dtype
Out[2]: dtype('int8')

In [3]: np.add.reduce(np.array([1, 2], dtype='i1')).dtype
Out[3]: dtype('int64')
```
(Just to be sure: python `int` does not promote to `int64`)

I'd be quite happy with an improvement to my suggestion, but let's not say something that is not actually the case - even if it looks like it, there is no account taken of the identity in casting.

p.s. @seberg - this long discussion is why I lamented that it was `np.add.reduce` that internally does the up-cast rather than leaving that for `np.sum`!

Reply:
Ok but you also raise an interesting example with `floor_divide`.

To summarize for now:
```pycon
>>> np.add.reduce([True, True], dtype=bool)
True  # `dtype` provided
>>> np.add.reduce([True, True], out=out)
array(2.)  # dtype of `out` when provided
>>> np.add.reduce([True, True])
2  # or result of casting with `ufunc.identity` (?)
>>> np.add.reduce([True, True], initial=.9)
2  # `initial` doesn't play any role
>>> 
>>> np.floor_divide.reduce([True, True])
1
>>> repr(np.floor_divide.identity)
'None'  # So when there is no identity, the rule doesn't apply
>>> np.result_type(np.True_, None)
dtype('float64')  # Not int, so the type doesn't even come from here
```

Maybe there's a special handling when there's no identity? I admit I'm a bit confused.

Reply:
The identify is just not used at all in casting - see https://github.com/numpy/numpy/blob/d3345bb1f5682cc54850708dfed213b1ae530a6f/numpy/_core/src/umath/ufunc_object.c#L2286-L2299